streaming/base/stream.py

```diff
@@ -173,8 +167,103 @@ def apply_default(self, default: Self) -> None:
             self.validate_hash = default.validate_hash or None
         if self._keep_zip is None:
             self.keep_zip = default.keep_zip
-        if self._keep_raw is None:
-            self.keep_raw = default.keep_raw
+            self.safe_keep_zip = self.keep_zip or self.remote in {None, self.local}
+
+    @classmethod
+    def validate_weights(cls, streams: Sequence[Self]) -> bool:
+        """Validate stream weights, returning whether relative or absolute weighting was used.
+
+        Args:
+            streams (Sequence[Stream]): Every stream comprising the dataset.
+
+        Returns:
+            bool: Whether streams are weighted relatively (proportionally).
+        """
+        # Validate sub-dataset weights ("proportion", "repeat", "samples", or none).
+        is_relative = hasattr(streams[0], 'proportion')
+        for stream_id, stream in enumerate(streams):
+            has_proportion = hasattr(stream, 'proportion')
+            has_repeat = hasattr(stream, 'repeat')
+            has_samples = hasattr(stream, 'samples')
+            if not (0 <= has_proportion + has_repeat + has_samples <= 1):
+                raise ValueError(f'Streams must provide at most one of "proportion", "repeat", ' +
+                                 f'or "samples" (error in stream {stream_id})')
+            if is_relative != has_proportion:
+                raise ValueError(f'Relative ("proportion") and absolute ("repeat", "samples", ' +
+                                 f'none) sub-dataset weights are incompatible with each other ' +
+                                 f'(error in stream {stream_id})')
+        return is_relative
+
+    @classmethod
+    def apply_weights(cls, streams: Sequence[Self], samples_per_stream: NDArray[np.int64],
+                      samples_per_epoch: Optional[int], seed: int) -> int:
+        """Given samples per stream, derive each stream's proportion/repeat/samples.
```

Comment:
note in docstring that this function modifies the streams objects to set the derived values

Reply:
done